Add JavaScript tutorial to the docs

perborgen opened the pull request:
This PR adds a tutorial on Image Object Detection using vanilla JavaScript. It's meant as the entry point for newbie devs who might struggle with the React/Next/Node examples. It requires no setup or build tools to run, simply drag&drop the file into the browser.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers.js/pr_271). All of your documentation changes will be reflected on that endpoint.

kungfooman commented on examples/javascript-object-detection/index.html:
```diff
@@ -0,0 +1,21 @@
+    <meta charset="UTF-8" />
+    <link rel="stylesheet" href="style.css" />
+    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+    <title>Web Worker Example</title>
```
I don't see any Worker creation

kungfooman commented:
1) The code is a mixture of spaces and tabs
2) Mixture of semicolons and lack thereof
3) Using `jsx` tags for JavaScript - any advantage over simply `js`?

But nice job :+1: :1st_place_medal:

perborgen commented:
Thanks @kungfooman!

I've changed it to 4 spaces for the indentation, and removed the jsx snippets. Not sure why they were there, probably just from Notion.

Also added semi colons :)

kungfooman commented:
I just love standalone examples, just pressing `.` while viewing your fork, GitHub editor opens, right click on the examples dir and download only one folder for some quick testing :+1: 

![image](https://github.com/xenova/transformers.js/assets/5236548/97a68f9f-474a-4d50-82a9-a6bc3aa4dec1)

That armor is a :cake: ... I wonder if we could also make a tutorial about fine-tuning :joy: Nice work :rocket:

xenova commented:
Thanks so much @perborgen! 😍 This is great! I'll review the whole thing soon. 🤗

xenova commented on docs/snippets/3_examples.snippet:
```diff
@@ -6,6 +6,7 @@ Want to jump straight in? Get started with one of our sample applications/templa
 | Doodle Dash       | Real-time sketch-recognition game (see [blog](https://huggingface.co/blog/ml-web-games)) | [link](https://github.com/xenova/doodle-dash) |
 | Code Playground   | In-browser code completion website | [link](./examples/code-completion/) |
 | Semantic Image Search | Search for images with text (Next.js + Supabase) | [link](./examples/semantic-image-search/) |
+| JavaScript        | In-browser image object detection | [link](./examples/javascript/) |
```
```suggestion
| Vanilla JavaScript        | In-browser object detection | [link](./examples/vanilla-javascript/) |
```

Perhaps we can rename it to "Vanilla JavaScript", and the thing which is special about this tutorial is that it doesn't use additional frameworks or bundlers.

NOTE: This means we need to rename it elsewhere too.

xenova commented on docs/source/tutorials/javascript.md:
```diff
@@ -0,0 +1,300 @@
+# Building a JavaScript Application
+
+In this tutorial, you’ll build a browser-based JavaScript application that detects objects in images using Transformers.js. The app requires no server, external libraries, or build tools. All you need is a code editor and a browser.
```
```suggestion
In this tutorial, you’ll build a simple JavaScript application that detects objects in images using Transformers.js. The app requires no server, external libraries, or build tools... all you need is a code editor and a browser!
```

Already mention "browser" in second line.

xenova commented on docs/source/tutorials/javascript.md:
```diff
@@ -0,0 +1,300 @@
+**************************Useful links:**************************
+
+- [Demo site](https://huggingface.co/spaces/Scrimba/javascript-object-detector)
+- [Source code walk through (scrim)](https://scrimba.com/scrim/co87f4ae7aea87a94a732f298)
```
```suggestion
- [Interactive code walk-through (scrim)](https://scrimba.com/scrim/co87f4ae7aea87a94a732f298)
```

I just also want to highlight how amazing this interactive video/code walk-though is! 🤗 
As discussed over slack, I'll suggest a few edits to the source code, but the functionality, explanation and demonstration you recorded is perfect! You mentioned it hopefully won't be too much effort to re-record, and after this revision I don't think we would need to make any more changes.

A small correction in the video: the model is 43 MB, not 20MB.

xenova commented on examples/javascript-object-detection/upload-icon.png:
I think we can move this to the transformers.js docs datasets repo on the HF hub (https://huggingface.co/datasets/Xenova/transformers.js-docs). -1 file in the editor too :)

xenova replied:
I see you've already uploaded some here: https://huggingface.co/datasets/Xenova/transformers.js-docs/discussions/1/files 🔥 great!

xenova commented on examples/javascript-object-detection/utils.js:
```diff
@@ -0,0 +1,53 @@
+}
+
+
+function getScaledCoordinates(box, img) {
```
I think we can greatly simplify this function. Perhaps using `percentage: true` can help with this too (one of the options to the detector):

![image](https://github.com/xenova/transformers.js/assets/26504141/73fa8990-9734-4d5b-bf9f-6606dc9c2bac)

(see [docs](https://huggingface.co/docs/transformers.js/api/pipelines#objectdetectionpipelinecallimages-options))

xenova commented on examples/javascript-object-detection/index.js:
```diff
@@ -0,0 +1,63 @@
+
+async function runModel(imageEl) {
+    statusParagraph.textContent = "Loading model...";
+    const detector = await pipeline('object-detection', 'Xenova/detr-resnet-50');
```
This creates a new pipeline every time we perform detection (and will keep allocating memory until we run out). I think we can simply move this to the top of the file instead. One downside is that it will start downloading on page load instead of lazily.

Maybe a better option would be to do `let detector;` at the top, and then `if (!detector) detector = await pipeline(...)`.

xenova commented on examples/javascript-object-detection/index.js:
```diff
@@ -0,0 +1,63 @@
+
+function renderBox(data, imageEl) {
+    const { box, label} = data;
+    const {xmax, xmin, ymax, ymin } = getScaledCoordinates(box, imageEl);
```
```suggestion
    const { xmax, xmin, ymax, ymin } = getScaledCoordinates(box, imageEl);
```

xenova commented on examples/javascript-object-detection/index.js:
```diff
@@ -0,0 +1,63 @@
+    const boxWidth = xmax - xmin;
+    const boxHeight = ymax - ymin;
+
+    // Draw the box
+    const boundingBoxElement = document.createElement("div");
+    boundingBoxElement.className = "bounding-box";
+    boundingBoxElement.style.border = `2px solid ${color}`;
+    boundingBoxElement.style.left = xmin + "px";
+    boundingBoxElement.style.top = ymin + "px";
+    boundingBoxElement.style.width = boxWidth + "px";
+    boundingBoxElement.style.height = boxHeight + "px";
+    imageContainer.appendChild(boundingBoxElement);
+
+    // Draw the label
+    let labelSpan = document.createElement('span');
+    labelSpan.textContent = label;
+    labelSpan.className = "bounding-box-label";
+    labelSpan.style.backgroundColor = color;
+    labelSpan.style.left = xmin + "px";
+    labelSpan.style.top = (ymin - 12) + "px";
+    imageContainer.appendChild(labelSpan);
```
Probably best to also move this to utils.js. I know it's not strictly a utility function, but it would be great to keep index.js as short as possible.

xenova reviewed (changes requested):
This is a game changer! I absolutely love the scrim too - very engaging and interactive.

I've left a few notes about things we can change, but what will probably be easiest (and quickest for me to review) is if you can grant me write access to your fork (https://github.com/perborgen/transformers.js).

Looking forward to getting this merged - thanks so much for your work!

perborgen commented:
Thanks for the review @xenova ! Just invited you to my fork so that you can make apply changes directly. Once I see how the final source code looks like, I'll record a new scrim :)

xenova commented:
@perborgen I've done my first pass of simplifying the app even more, so we don't need an extra utils.js file. Some TODOs:
- [ ] Update docs
- [ ] Rename "JavaScript" -> "Vanilla JS"
- [ ] Try avoid using separate span for labels.

I think it's a lot cleaner now, and we can continue discussing what more can be done.

perborgen reviewed (commented):
This is so much better, thanks @xenova !

xenova commented on docs/source/tutorials/vanilla-js.md:
````diff
@@ -0,0 +1,281 @@
+}
+```
+
+If you drag and drop the `index.html` file into a browser, you should see the following:
````
I think that adding this line might be a bit confusing later on, because when they refresh it will show a [CORS](https://stackoverflow.com/questions/52919331/access-to-script-at-from-origin-null-has-been-blocked-by-cors-policy) error after they try adding JS.

> Access to script at 'file:///.../index.js' from origin 'null' has been blocked by CORS policy: Cross origin requests are only supported for protocol schemes: http, data, isolated-app, chrome-extension, chrome, https, chrome-untrusted.

Ideally we could link to some online playground for this. Of course, using a scrim would be great, but it requires a sign up before starting to use the editor (unlike jsfiddle, codepen, or codesandbox). WDYT?

perborgen replied:
Ah, got it. You actually don't need an account to start editing a scrim, so we can point to a scrim and let them run the code. Does that sound good?

xenova replied:
Sure that works! 😇 In that case, we can just link to the ready-made scrim (unless you think it's better to link to an empty one?)

perborgen replied:
How about I link to a scrim which mirrors the code at that exact point in the tutorial?

xenova replied:
Is there a way to have "chapters" or "versions" of a scrim? In that way, we can link to a version of the scrim that is not yet complete, but the user can click a button or continue the video/walkthrough to get the rest of it?

perborgen replied:
We can link to specific times in the timeline, so that they can continue watching from that point. For example, here's a link to the `runModel() ` explanation of the old scrim: https://scrimba.com/scrim/co87f4ae7aea87a94a732f298?a=7877.2.3.L2_31-2_45

perborgen replied:
I updated the text to simply not talk about them running the code themselves. That way we dodge the whole CORS issue in that sentence.

Have added a mention about the need for a server in the first paragraph though.

xenova replied:
Sounds good! 👍

xenova commented:
I've just been proof-reading and improving some of the sections (mainly just trying to keep some terms consistent across the docs). I'll continue with step 4 and the rest tomorrow.

xenova reviewed (approved):
Good to go!
